presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeMatchingChecker.java

```diff
@@ -0,0 +1,169 @@
+
+    @Override
+    public void validate(Session session, Metadata metadata, SqlParser sqlParser, Map<Symbol, Type> types, PlanNode plan)
+    {
+        plan.accept(new Visitor(session, metadata, sqlParser, types), null);
+    }
+
+    private static class Visitor
+            extends SimplePlanVisitor<Void>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final SqlParser sqlParser;
+        private final Map<Symbol, Type> types;
+
+        public Visitor(Session session, Metadata metadata, SqlParser sqlParser, Map<Symbol, Type> types)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+            this.types = requireNonNull(types, "types is null");
+        }
+
+        @Override
+        public Void visitAggregation(AggregationNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            AggregationNode.Step step = node.getStep();
+
+            // Do checking only when the step is SINGLE since FINAL and PARTIAL aggregation might have intermediate representation types
+            if (step.equals(AggregationNode.Step.SINGLE)) {
+                checkSignatureTypeMatching(node.getFunctions());
+                checkFunctionCallTypeMatching(node.getAggregations());
+            }
+            else if (step.equals(AggregationNode.Step.FINAL)) {
+                checkSignatureTypeMatching(node.getFunctions());
+            }
+
+            return null;
+        }
+
+        @Override
+        public Void visitWindow(WindowNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            checkSignatureTypeMatching(node.getSignatures());
+            checkFunctionCallTypeMatching(node.getWindowFunctions());
+
+            return null;
+        }
+
+        @Override
+        public Void visitProject(ProjectNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            checkExpressionTypeMatching(node.getAssignments());
+
+            return null;
+        }
+
+        @Override
+        public Void visitUnion(UnionNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            ListMultimap<Symbol, Symbol> symbolMapping = node.getSymbolMapping();
+            for (Symbol keySymbol : symbolMapping.keySet()) {
+                List<Symbol> valueSymbols = symbolMapping.get(keySymbol);
+                Type expectedType = types.get(keySymbol);
+                for (Symbol valueSymbol : valueSymbols) {
+                    verifyTypeSignature(keySymbol, expectedType.getTypeSignature(), types.get(valueSymbol).getTypeSignature());
+                }
+            }
+
+            return null;
+        }
+
+        private void checkSignatureTypeMatching(Map<Symbol, Signature> functions)
```

Comment:
I suggest simply calling this checkFunctionSignature, within the context of this class it's clear what is being checked. Same for checkFunctionCall